README.md
[![Latest Stable Version](http://poser.pugx.org/nguyenanhung/string-helper/v)](https://packagist.org/packages/nguyenanhung/string-helper) [![Total Downloads](http://poser.pugx.org/nguyenanhung/string-helper/downloads)](https://packagist.org/packages/nguyenanhung/string-helper) [![Latest Unstable Version](http://poser.pugx.org/nguyenanhung/string-helper/v/unstable)](https://packagist.org/packages/nguyenanhung/string-helper) [![License](http://poser.pugx.org/nguyenanhung/string-helper/license)](https://packagist.org/packages/nguyenanhung/string-helper) [![PHP Version Require](http://poser.pugx.org/nguyenanhung/string-helper/require/php)](https://packagist.org/packages/nguyenanhung/string-helper)

# String Helper

String Helper, basic, simple and lightweight

## Version

- [x] V1.x support all PHP version `>=5.6`
- [x] V2.x support all PHP version `>=7.0`

## Contact & Support

If any question & request, please contact following information

| Name        | Email                | Skype            | Facebook      |
| ----------- | -------------------- | ---------------- | ------------- |
| Hung Nguyen | dev@nguyenanhung.com | nguyenanhung5891 | @nguyenanhung |

From Vietnam with Love <3


CHANGELOG.md
# CHANGELOG

## [1.0.0] - 2021/09/22

### Added

- [x] Init Project

## Contact & Support

If any question & request, please contact following information

| Name        | Email                | Skype            | Facebook      |
| ----------- | -------------------- | ---------------- | ------------- |
| Hung Nguyen | dev@nguyenanhung.com | nguyenanhung5891 | @nguyenanhung |


test/test.php
<?php
/**
 * Project string-helper
 * Created by PhpStorm
 * User: 713uk13m <dev@nguyenanhung.com>
 * Copyright: 713uk13m <dev@nguyenanhung.com>
 * Date: 09/22/2021
 * Time: 14:15
 */

src/Utf8.php
<?php
/**
 * Project string-helper
 * Created by PhpStorm
 * User: 713uk13m <dev@nguyenanhung.com>
 * Copyright: 713uk13m <dev@nguyenanhung.com>
 * Date: 09/22/2021
 * Time: 18:22
 */

namespace nguyenanhung\Libraries\String;

if (!class_exists('nguyenanhung\Libraries\String\Utf8')) {
    /**
     * Utf8 Class
     *
     * Provides support for UTF-8 environments
     *
     * @package        CodeIgniter
     * @subpackage     Libraries
     * @category       UTF-8
     * @author         EllisLab Dev Team
     * @link           https://codeigniter.com/user_guide/libraries/utf8.html
     */
    class Utf8
    {
        /**
         * Utf8 constructor.
         * Determines if UTF-8 support is to be enabled.
         */
        public function __construct()
        {
            $charset = 'UTF-8';
            if (extension_loaded('mbstring')) {
                define('MB_ENABLED', true);
                // mbstring.internal_encoding is deprecated starting with PHP 5.6
                // and it's usage triggers E_DEPRECATED messages.
                @ini_set('mbstring.internal_encoding', $charset);
                // This is required for mb_convert_encoding() to strip invalid characters.
                // That's utilized by CI_Utf8, but it's also done for consistency with iconv.
                mb_substitute_character('none');
            } else {
                define('MB_ENABLED', false);
            }
            // There's an ICONV_IMPL constant, but the PHP manual says that using
            // iconv's predefined constants is "strongly discouraged".
            if (extension_loaded('iconv')) {
                define('ICONV_ENABLED', true);
                // iconv.internal_encoding is deprecated starting with PHP 5.6
                // and it's usage triggers E_DEPRECATED messages.
                @ini_set('iconv.internal_encoding', $charset);
            } else {
                define('ICONV_ENABLED', false);
            }
            if (
                defined('PREG_BAD_UTF8_ERROR')                // PCRE must support UTF-8
                && (ICONV_ENABLED === true or MB_ENABLED === true)    // iconv or mbstring must be installed
                && strtoupper($charset) === 'UTF-8'    // Application charset must be UTF-8
            ) {
                define('UTF8_ENABLED', true);
            } else {
                define('UTF8_ENABLED', false);
            }
        }

        // --------------------------------------------------------------------

        /**
         * Clean UTF-8 strings
         *
         * Ensures strings contain only valid UTF-8 characters.
         *
         * @param string $str String to clean
         *
         * @return    string
         */
        public function clean_string(string $str): string
        {
            if ($this->is_ascii($str) === false) {
                if (MB_ENABLED) {
                    $str = mb_convert_encoding($str, 'UTF-8', 'UTF-8');
                } elseif (ICONV_ENABLED) {
                    $str = iconv('UTF-8', 'UTF-8//IGNORE', $str);
                }
            }

            return $str;
        }

        // --------------------------------------------------------------------

        /**
         * Remove ASCII control characters
         *
         * Removes all ASCII control characters except horizontal tabs,
         * line feeds, and carriage returns, as all others can cause
         * problems in XML.
         *
         * @param string $str String to clean
         *
         * @return    string
         */
        public function safe_ascii_for_xml(string $str): string
        {
            return remove_invisible_characters_string($str, false);
        }

        // --------------------------------------------------------------------

        /**
         * Convert to UTF-8
         *
         * Attempts to convert a string to UTF-8.
         *
         * @param string $str      Input string
         * @param string $encoding Input encoding
         *
         * @return    string    $str encoded in UTF-8 or FALSE on failure
         */
        public function convert_to_utf8(string $str, string $encoding)
        {
            if (MB_ENABLED) {
                return mb_convert_encoding($str, 'UTF-8', $encoding);
            }

            if (ICONV_ENABLED) {
                return @iconv($encoding, 'UTF-8', $str);
            }

            return false;
        }

        // --------------------------------------------------------------------

        /**
         * Is ASCII?
         *
         * Tests if a string is standard 7-bit ASCII or not.
         *
         * @param string $str String to check
         *
         * @return    bool
         */
        public function is_ascii(string $str): bool
        {
            return (preg_match('/[^\x00-\x7F]/S', $str) === 0);
        }
    }
}



src/config/string.php
<?php
/**
 * Project helpers.
 * Created by PhpStorm.
 * User: 713uk13m <dev@nguyenanhung.com>
 * Date: 11/6/18
 * Time: 09:07
 */
return [
    'separator'     => '-',
    'vn_array'      => ["à", "á", "ạ", "ả", "ã", "ầ", "ấ", "ậ", "ẩ", "ẫ", "ă", "â", "ằ", "ắ", "ặ", "ẳ", "ẵ", "ă", "è", "é", "ẹ", "ẻ", "ẽ", "ề", "ế", "ệ", "ể", "ễ", "ê", "ì", "í", "ị", "ỉ", "ĩ", "ò", "ó", "ọ", "ỏ", "õ", "ồ", "ố", "ộ", "ổ", "ỗ", "ô",
                        "ờ", "ớ", "ợ", "ở", "ỡ", "ơ", "ù", "ú", "ụ", "ủ", "ũ", "ừ", "ứ", "ự", "ử", "ữ", "ư", "ỳ", "ý", "ỵ", "ỷ", "ỹ", "đ", "A", "À", "Á", "Ạ", "Ả", "Ã", "Ầ", "Ấ", "Ậ", "Ẩ", "Ẫ", "Ă", "Â", "Ằ", "Ắ", "Ặ", "Ẳ", "Ẵ", "Ă", "È", "É", "Ẹ",
                        "Ẻ", "Ẽ", "E", "Ề", "Ế", "Ệ", "Ể", "Ễ", "Ê", "I", "Ì", "Í", "Ị", "Ỉ", "Ĩ", "O", "Ò", "Ó", "Ọ", "Ỏ", "Õ", "Ồ", "Ố", "Ộ", "Ổ", "Ỗ", "Ô", "Ờ", "Ớ", "Ợ", "Ở", "Ỡ", "Ơ", "Ù", "Ú", "Ụ", "Ủ", "Ũ", "U", "Ừ", "Ứ", "Ự", "Ử", "Ữ", "Ư",
                        "Ỳ", "Ý", "Ỵ", "Ỷ", "Ỹ", "Y", "Đ", "B", "C", "D", "F", "G", "H", "I", "J", "K", "L", "M", "N", "P", "Q", "R", "S", "T", "V", "X", "Y", "Z", "W"],
    'en_array'      => ["a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "e", "e", "e", "e", "e", "e", "e", "e", "e", "e", "e", "i", "i", "i", "i", "i", "o", "o", "o", "o", "o", "o", "o", "o", "o", "o", "o",
                        "o", "o", "o", "o", "o", "o", "u", "u", "u", "u", "u", "u", "u", "u", "u", "u", "u", "y", "y", "y", "y", "y", "d", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "e", "e", "e",
                        "e", "e", "e", "e", "e", "e", "e", "e", "e", "i", "i", "i", "i", "i", "i", "o", "o", "o", "o", "o", "o", "o", "o", "o", "o", "o", "o", "o", "o", "o", "o", "o", "o", "u", "u", "u", "u", "u", "u", "u", "u", "u", "u", "u", "u",
                        "y", "y", "y", "y", "y", "y", "d", "b", "c", "d", "f", "g", "h", "i", "j", "k", "l", "m", "n", "p", "q", "r", "s", "t", "v", "x", "y", "z", "w"],
    'special_array' => [',', ';', '\'', '"', '(', ')', '.', ':', '…', '[', ']', '|', '\\', '?', "/", "!", "@", "#", "$", "^", "&", "*", "+", "=", "<", ">", "–", '™', '®', '%', '“', '”', '’', '‘'],
    'ascii_array'   => ['&#225;', '&#224;', '&#7843;', '&#227;', '&#7841;', '&#259;', '&#7855;', '&#7857;', '&#7859;', '&#7861;', '&#7863;', '&#226;', '&#7845;', '&#7847;', '&#7849;', '&#7851;', '&#7853;', '&#273;', '&#233;', '&#232;', '&#7867;',
                        '&#7869;', '&#7865;', '&#234;', '&#7871;', '&#7873;', '&#7875;', '&#7877;', '&#7879;', '&#237;', '&#236;', '&#7881;', '&#297;', '&#7883;', '&#243;', '&#242;', '&#7887;', '&#245;', '&#7885;', '&#244;', '&#7889;', '&#7891;',
                        '&#7893;', '&#7895;', '&#7897;', '&#417;', '&#7899;', '&#7901;', '&#7903;', '&#7905;', '&#7907;', '&#250;', '&#249;', '&#7911;', '&#361;', '&#7909;', '&#432;', '&#7913;', '&#7915;', '&#7917;', '&#7919;', '&#7921;', '&#253;',
                        '&#7923;', '&#7927;', '&#7929;', '&#7925;', '&#193;', '&#192;', '&#7842;', '&#195;', '&#7840;', '&#258;', '&#7854;', '&#7856;', '&#7858;', '&#7860;', '&#7862;', '&#194;', '&#7844;', '&#7846;', '&#7848;', '&#7850;', '&#7852;',
                        '&#272;', '&#201;', '&#200;', '&#7866;', '&#7868;', '&#7864;', '&#202;', '&#7870;', '&#7872;', '&#7874;', '&#7876;', '&#7878;', '&#205;', '&#204;', '&#7880;', '&#296;', '&#7882;', '&#211;', '&#210;', '&#7886;', '&#213;',
                        '&#7884;', '&#212;', '&#7888;', '&#7890;', '&#7892;', '&#7894;', '&#7896;', '&#416;', '&#7898;', '&#7900;', '&#7902;', '&#7904;', '&#7906;', '&#218;', '&#217;', '&#7910;', '&#360;', '&#7908;', '&#431;', '&#7912;', '&#7914;',
                        '&#7916;', '&#7918;', '&#7920;', '&#221;', '&#7922;', '&#7926;', '&#7928;', '&#7924;'],
    'utf8_array'    => ['á', 'à', 'ả', 'ã', 'ạ', 'ă', 'ắ', 'ằ', 'ẳ', 'ẵ', 'ặ', 'â', 'ấ', 'ầ', 'ẩ', 'ẫ', 'ậ', 'đ', 'é', 'è', 'ẻ', 'ẽ', 'ẹ', 'ê', 'ế', 'ề', 'ể', 'ễ', 'ệ', 'í', 'ì', 'ỉ', 'ĩ', 'ị', 'ó', 'ò', 'ỏ', 'õ', 'ọ', 'ó ', 'ô', 'ố', 'ồ', 'ổ', 'ỗ',
                        'ộ', 'ơ', 'ớ', 'ờ', 'ở', 'ỡ', 'ợ', 'ú', 'ù', 'ủ', 'ũ', 'ụ', 'u', 'ư', 'ứ', 'ừ', 'ử', 'ữ', 'ự', 'ý', 'ỳ', 'ỷ', 'ỹ', 'ỵ', 'Á', 'À', 'Ả', 'Ã', 'Ạ', 'Ă', 'Ắ', 'Ằ', 'Ẳ', 'Ẵ', 'Ặ', 'Â', 'Ấ', 'Ầ', 'Ẩ', 'Ẫ', 'Ậ', 'Đ', 'É', 'È', 'Ẻ',
                        'Ẽ', 'Ẹ', 'Ê', 'Ế', 'Ề', 'Ể', 'Ễ', 'Ệ', 'Í', 'Ì', 'Ỉ', 'Ĩ', 'Ị', 'Ó', 'Ò', 'Ỏ', 'Õ', 'Ọ', 'Ô', 'Ố', 'Ồ', 'Ổ', 'Ỗ', 'Ộ', 'Ơ', 'Ớ', 'Ờ', 'Ở', 'Ỡ', 'Ợ', 'Ú', 'Ù', 'Ủ', 'Ũ', 'Ụ', 'Ư', 'Ứ', 'Ừ', 'Ử', 'Ữ', 'Ự', 'Ý', 'Ỳ', 'Ỷ', 'Ỹ',
                        'Ỵ'],
    'normal_array'  => ['a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'd', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'i', 'i', 'i', 'i', 'i', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o',
                        'o', 'o', 'o', 'o', 'o', 'o', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'y', 'y', 'y', 'y', 'y', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'D', 'E', 'E', 'E', 'E', 'E',
                        'E', 'E', 'E', 'E', 'E', 'E', 'I', 'I', 'I', 'I', 'I', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'U', 'U', 'U', 'U', 'U', 'U', 'U', 'U', 'U', 'U', 'U', 'Y', 'Y', 'Y', 'Y', 'Y']
];


src/DataRepository.php
<?php
/**
 * Project string-helper
 * Created by PhpStorm
 * User: 713uk13m <dev@nguyenanhung.com>
 * Copyright: 713uk13m <dev@nguyenanhung.com>
 * Date: 09/22/2021
 * Time: 17:28
 */

namespace nguyenanhung\Libraries\String;

if (!class_exists('nguyenanhung\Libraries\String\DataRepository')) {
    /**
     * Class DataRepository
     *
     * @package   nguyenanhung\Libraries\String
     * @author    713uk13m <dev@nguyenanhung.com>
     * @copyright 713uk13m <dev@nguyenanhung.com>
     */
    class DataRepository
    {
        const CONFIG_PATH = 'config';
        const CONFIG_EXT  = '.php';

        /**
         * Hàm lấy nội dung config được quy định trong thư mục config
         *
         * @author: 713uk13m <dev@nguyenanhung.com>
         * @time  : 9/28/18 14:47
         *
         * @param string $configName Tên file config
         *
         * @return array|mixed
         */
        public static function getData(string $configName)
        {
            $path = __DIR__ . DIRECTORY_SEPARATOR . self::CONFIG_PATH . DIRECTORY_SEPARATOR . $configName . self::CONFIG_EXT;
            if (is_file($path) && file_exists($path)) {
                return require $path;
            }

            return [];
        }

        /**
         * Hàm lấy nội dung Data từ 1 file bất kỳ trong hệ thống
         *
         * @author: 713uk13m <dev@nguyenanhung.com>
         * @time  : 10/17/18 09:25
         *
         * @param string $filename Đường dẫn file config
         *
         * @return array|mixed
         */
        public static function getDataContent(string $filename)
        {
            if (is_file($filename) && file_exists($filename)) {
                return require $filename;
            }

            return [];
        }
    }
}


helpers/ci_string_helper.php
<?php
/**
 * Project string-helper
 * Created by PhpStorm
 * User: 713uk13m <dev@nguyenanhung.com>
 * Copyright: 713uk13m <dev@nguyenanhung.com>
 * Date: 09/22/2021
 * Time: 17:42
 */
/**
 * CodeIgniter
 *
 * An open source application development framework for PHP
 *
 * This content is released under the MIT License (MIT)
 *
 * Copyright (c) 2014 - 2018, British Columbia Institute of Technology
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * @package      CodeIgniter
 * @author       EllisLab Dev Team
 * @copyright    Copyright (c) 2008 - 2014, EllisLab, Inc. (https://ellislab.com/)
 * @copyright    Copyright (c) 2014 - 2018, British Columbia Institute of Technology (http://bcit.ca/)
 * @license      http://opensource.org/licenses/MIT	MIT License
 * @link         https://codeigniter.com
 * @since        Version 1.0.0
 * @filesource
 */
/**
 * CodeIgniter String Helpers
 *
 * @package        CodeIgniter
 * @subpackage     Helpers
 * @category       Helpers
 * @author         EllisLab Dev Team
 * @link           https://codeigniter.com/user_guide/helpers/string_helper.html
 */

// ------------------------------------------------------------------------

if (!function_exists('trim_slashes')) {
    /**
     * Trim Slashes
     *
     * Removes any leading/trailing slashes from a string:
     *
     * /this/that/theother/
     *
     * becomes:
     *
     * this/that/theother
     *
     * @todo          Remove in version 3.1+.
     * @deprecated    3.0.0    This is just an alias for PHP's native trim()
     *
     * @param string
     *
     * @return    string
     */
    function trim_slashes($str): string
    {
        return trim($str, '/');
    }
}

// ------------------------------------------------------------------------

if (!function_exists('strip_slashes')) {
    /**
     * Strip Slashes
     *
     * Removes slashes contained in a string or in an array
     *
     * @param mixed    string or array
     *
     * @return    array|string    string or array
     */
    function strip_slashes($str)
    {
        if (!is_array($str)) {
            return stripslashes($str);
        }

        foreach ($str as $key => $val) {
            $str[$key] = strip_slashes($val);
        }

        return $str;
    }
}

// ------------------------------------------------------------------------

if (!function_exists('strip_quotes')) {
    /**
     * Strip Quotes
     *
     * Removes single and double quotes from a string
     *
     * @param string
     *
     * @return    string
     */
    function strip_quotes($str): string
    {
        return str_replace(['"', "'"], '', $str);
    }
}

// ------------------------------------------------------------------------

if (!function_exists('quotes_to_entities')) {
    /**
     * Quotes to Entities
     *
     * Converts single and double quotes to entities
     *
     * @param string
     *
     * @return    string
     */
    function quotes_to_entities($str): string
    {
        return str_replace(["\'", "\"", "'", '"'], ["&#39;", "&quot;", "&#39;", "&quot;"], $str);
    }
}

// ------------------------------------------------------------------------

if (!function_exists('reduce_double_slashes')) {
    /**
     * Reduce Double Slashes
     *
     * Converts double slashes in a string to a single slash,
     * except those found in http://
     *
     * http://www.some-site.com//index.php
     *
     * becomes:
     *
     * http://www.some-site.com/index.php
     *
     * @param string
     *
     * @return    string
     */
    function reduce_double_slashes($str): string
    {
        return preg_replace('#(^|[^:])//+#', '\\1/', $str);
    }
}

// ------------------------------------------------------------------------

if (!function_exists('reduce_multiples')) {
    /**
     * Reduce Multiples
     *
     * Reduces multiple instances of a particular character.  Example:
     *
     * Fred, Bill,, Joe, Jimmy
     *
     * becomes:
     *
     * Fred, Bill, Joe, Jimmy
     *
     * @param string
     * @param string    the character you wish to reduce
     * @param bool    TRUE/FALSE - whether to trim the character from the beginning/end
     *
     * @return    string
     */
    function reduce_multiples($str, $character = ',', $trim = false): string
    {
        $str = preg_replace('#' . preg_quote($character, '#') . '{2,}#', $character, $str);

        return ($trim === true) ? trim($str, $character) : $str;
    }
}

// ------------------------------------------------------------------------

if (!function_exists('random_string')) {
    /**
     * Create a "Random" String
     *
     * @param string    type of random string.  basic, alpha, alnum, numeric, nozero, unique, md5, encrypt and sha1
     * @param int    number of characters
     *
     * @return    string
     */
    function random_string($type = 'alnum', $len = 8): string
    {
        switch ($type) {
            case 'basic':
                return mt_rand();
            case 'alnum':
            case 'numeric':
            case 'nozero':
            case 'alpha':
                switch ($type) {
                    case 'alpha':
                        $pool = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
                        break;
                    case 'numeric':
                        $pool = '0123456789';
                        break;
                    case 'nozero':
                        $pool = '123456789';
                        break;
                    default:
                        $pool = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
                        break;
                }

                return substr(str_shuffle(str_repeat($pool, ceil($len / strlen($pool)))), 0, $len);
            case 'unique': // todo: remove in 3.1+
            case 'md5':
                return md5(uniqid(mt_rand(), true));
            case 'encrypt': // todo: remove in 3.1+
            case 'sha1':
                return sha1(uniqid(mt_rand(), true));
            default:
                return md5(mt_rand());
        }
    }
}

// ------------------------------------------------------------------------

if (!function_exists('increment_string')) {
    /**
     * Add's _1 to a string or increment the ending number to allow _2, _3, etc
     *
     * @param string     $str       required
     * @param string     $separator What should the duplicate number be appended with
     * @param string|int $first     Which number should be used for the first dupe increment
     *
     * @return    string
     */
    function increment_string(string $str, string $separator = '_', $first = 1): string
    {
        preg_match('/(.+)' . preg_quote($separator, '/') . '([0-9]+)$/', $str, $match);

        return isset($match[2]) ? $match[1] . $separator . ($match[2] + 1) : $str . $separator . $first;
    }
}

// ------------------------------------------------------------------------

if (!function_exists('alternator')) {
    /**
     * Alternator
     *
     * Allows strings to be alternated. See docs...
     *
     * @param string (as many parameters as needed)
     *
     * @return    string
     */
    function alternator(): string
    {
        static $i;

        if (func_num_args() === 0) {
            $i = 0;

            return '';
        }

        $args = func_get_args();

        return $args[($i++ % count($args))];
    }
}

// ------------------------------------------------------------------------

if (!function_exists('repeater')) {
    /**
     * Repeater function
     *
     * @todo          Remove in version 3.1+.
     * @deprecated    3.0.0    This is just an alias for PHP's native str_repeat()
     *
     * @param string $data String to repeat
     * @param int    $num  Number of repeats
     *
     * @return    string
     */
    function repeater(string $data, int $num = 1): string
    {
        return ($num > 0) ? str_repeat($data, $num) : '';
    }
}

// ------------------------------------------------------------------------

if (!function_exists('remove_invisible_characters_string')) {
    /**
     * Remove Invisible Characters
     *
     * This prevents sandwiching null characters
     * between ascii characters, like Java\0script.
     *
     * @param string
     * @param bool
     *
     * @return    string
     */
    function remove_invisible_characters_string($str, $url_encoded = true): string
    {
        $nonDisplay = array();
        // every control character except newline (dec 10),
        // carriage return (dec 13) and horizontal tab (dec 09)
        if ($url_encoded) {
            $nonDisplay[] = '/%0[0-8bcef]/i';    // url encoded 00-08, 11, 12, 14, 15
            $nonDisplay[] = '/%1[0-9a-f]/i';    // url encoded 16-31
            $nonDisplay[] = '/%7f/i';    // url encoded 127
        }
        $nonDisplay[] = '/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]+/S';    // 00-08, 11, 12, 14-31, 127
        do {
            $str = preg_replace($nonDisplay, '', $str, -1, $count);
        }
        while ($count);

        return $str;
    }
}